### PR TITLE
DEVENGAGE-569_java - Adding config and logging to the Java SDK

### PR DESCRIPTION
**Pull request by jacob-shaw-genesys**

### CODE REVIEW

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 1]
```diff
@@ -168,22 +168,121 @@ Specify the connector in the builder:
...
+`LogFormat` values:
+1. JSON
+2. Text - default
+
+By default, the request and response bodies are not logged because these can contain PII. Be mindful of this data if choosing to log it.  
+To log to a file, provide a `LogFilePath` value. SDK users are responsible for the rotation of the log file.
+
+Example logging configuration:
+```{"language":"java"}
+ApiClient.LoggingConfiguration = new ApiClient.LoggingConfiguration();
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:208]: I'm getting an error here. I guess it should be `ApiClient.LoggingConfiguration loggingConfiguration = new ApiClient.LoggingConfiguration();`
Also show that you add it to the ApiClient with `withLoggingConfiguration`

File: resources/sdk/purecloudjava/templates/Logger.mustache [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        return null;
+    }
+
+    public enum LogFormat {
+        JSON,
+        Text
+    }
+    
+    public enum LogLevel {
+        LNone(4),
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/Logger.mustache:228]: The levels don't seem to be correct. I'm seeing trace logs when it's set to debug. Debug should have debug and error.
I can't see any debug messages when it's set to trace, Trace should have trace, debug and error.
Can't see any error logs regardless of level.

File: resources/sdk/purecloudjava/templates/ApiClient.mustache [hunk 3]
```diff
@@ -672,14 +771,21 @@ public class ApiClient implements AutoCloseable {
             Retry retry = new Retry(retryConfiguration);
             do {
                 connectorResponse = connector.invoke(connectorRequest);
+                // copy unmodifiable maps to make modifiable
+                Map<String,String> requestHeaderCopy = new HashMap<>(request.getHeaderParams());
+                Map<String,String> responseHeaderCopy = new HashMap<>(connectorResponse.getHeaders());
+                logger.debug(request.getMethod(), request.getPath(), (Object)request.getBody(), connectorResponse.getStatusCode(), requestHeaderCopy);
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/ApiClient.mustache:777]: The only request header I'm seeing is the "Authorization: [REDACTED]" header. There should be a few more headers in the request. Check that your source for the headers is correct

File: resources/sdk/purecloudjava/templates/ApiClient.mustache [hunk 4]
```diff
@@ -956,6 +1075,21 @@ public class ApiClient implements AutoCloseable {
             return this;
         }
 
+        public Builder withLogger(Logger logger) {
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/ApiClient.mustache:1078]: Probably best to leave this out and just let them use withLoggingConfiguration, unless you can think of a good reason for it.

File: resources/sdk/purecloudjava/templates/ApiClient.mustache [hunk 5]
```diff
@@ -1178,30 +1313,107 @@ public class ApiClient implements AutoCloseable {
...
+            this.retryMax = retryMax;
+        }
     }
 
+    public static class LoggingConfiguration {
+        private String logLevel = "none";
+        private String logFilePath = null;
+        private String logFormat = "text";
+        private Boolean logToConsole = true;
+        private Boolean logRequestBody = true;
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/ApiClient.mustache:1330]: This should default to false

File: resources/sdk/purecloudjava/tests/.idea/misc.xml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/tests/.idea/misc.xml:1]: Should leave this .idea directory out of the commit

File: resources/sdk/purecloudjava/tests/tests.iml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/tests/tests.iml:1]: Not sure if this should be added

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 3 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache:777]: Also, if possible try to log the full URL of the path. e.g. https://api.mypurecloud.com/api/v2/users/me instead of /api/v2/users/me

File: resources/sdk/purecloudjava/templates/Configuration.mustache [hunk 8]
```diff
@@ -20,5 +36,245 @@ public class Configuration {
...
+    autoReloadConfig = defaultApiClient.getAutoReloadConfig();
+    logger = defaultApiClient.getLogger();
+    configFilePath = defaultApiClient.getConfigFilePath();
+    ApiClient.LoggingConfiguration loggingConfiguration = defaultApiClient.getLoggingConfiguration();
+    Configuration configuration = new Configuration();
+    configFileHash = configuration.getConfigHash();
+    configuration.applyConfigFromFile();
+    if (autoReloadConfig) {
+      Runnable runnable = () -> configuration.runConfigChecker();
+      Thread configCheckerThread = new Thread(runnable);
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/Configuration.mustache:48]: This should be set to a daemon thread so it doesn't keep the application running when all other threads have finished

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 3 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache:777]: Also, I tried making a post request and the request body wasn't logged despite logRequestBody being set to true. I used a debugger and requestBodyToString returned null because neither of the conditions checked in it was met.

Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache:777]: The request was made using:
```
CreateUser createUser = new CreateUser();
createUser.setName("name");
usersApi.postUsers(createUser);
```

File: resources/sdk/purecloudjava/templates/Configuration.mustache [hunk 9]
```diff
@@ -20,5 +36,245 @@ public class Configuration {
...
+        return -1;
+      }
+    }
+
+    public String getIniString(String section, String key) {
+      try {
+        Ini.Section sectionData = iniData.get(section);
+        return sectionData.get(key).trim();
+      } catch (Exception e) {
+        return "";
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/Configuration.mustache:251]: I tried using an INI file with the same contents as the README and it wasn't applying the config. I used a debugger on this method when it was called with `getIniString("logging", "log_level")` and there was a NullPointerException being thrown

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 10]
```diff
@@ -168,22 +168,121 @@ Specify the connector in the builder:
...
+loggingConfiguration.setLogLevel("trace");
+loggingConfiguration.setLogFormat("json");
+loggingConfiguration.setLogRequestBody(true);
+loggingConfiguration.setLogResponseBody(true);
+loggingConfiguration.setLogToConsole(true);
+loggingConfiguration.setLogFilePath("src/main/resources/javasdk.log");
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:214]: Choose a full path for this. e.g. /var/log/javasdk.log
Same thing with the config path on 227

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 11]
```diff
@@ -168,22 +168,121 @@ Specify the connector in the builder:
...
+
+A number of configuration parameters can be applied using a configuration file. There are two sources for this file:
+
+1. The SDK will look for `%HOMEDRIVE%%HOMEPATH%\.genesyscloudjava\config` on Windows, or `$HOME/.genesyscloudjava/config` on Unix.
+2. Provide a valid file path to the ApiClient
+
+Example setting the configuration file:
+
+```{"language": "java"}
+ApiClient apiClient = ApiClient.Builder.standard().withConfigFilePath("src/main/resources/logconfig.json").build();
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:227]: Might be nicer to just show the `withConfigFilePath("/path/config")` here as is done in the "Setting the access token" section. Same goes for line 234

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 12]
```diff
@@ -168,22 +168,121 @@ Specify the connector in the builder:
...
+```
+INI and JSON formats are supported. See below for examples of configuration values in both formats:
+
+INI:
+```{"language":"ini"}
+[logging]
+log_level = trace
+log_format = text
+log_to_console = false
+log_file_path = src/main/resources/javasdk.log
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:244]: Use a full path here and in the JSON config example as well.

File: resources/sdk/purecloudjava/templates/pom.mustache [hunk 13]
```diff
@@ -223,11 +223,21 @@
...
     <version>2.2.12</version>
 </dependency>
- <!-- Logging: SLF4J -->
     <dependency>
-      <groupId>org.slf4j</groupId>
-      <artifactId>slf4j-api</artifactId>
-      <version>${slf4j-version}</version>
+      <groupId>org.ini4j</groupId>
+      <artifactId>ini4j</artifactId>
+      <version>0.5.4</version>
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/pom.mustache:229]: You can put the versions for the added dependencies down below in the section with all the other dependency versions. So this one would become ${ini4j-version} for example.

File: resources/sdk/purecloudjava/templates/Configuration.mustache [hunk 14]
```diff
@@ -20,5 +36,245 @@ public class Configuration {
...
+
+  public void runConfigChecker() {
+    while (true) {
+        try {
+          Thread.sleep(30 * 1000);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          break;
+        }
+        if (!autoReloadConfig) {
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/Configuration.mustache:139]: Seems like this is never being set to false using a config file. I tried having it set to false before starting the application and while it was running but it stayed true. That was with INI and JSON

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 3 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache:777]: Also, I tried calling `notificationsApi.deleteNotificationsChannelSubscriptions("testchannelid");` and it logged the URL as `/api/v2/notifications/channels/{channelId}/subscriptions` instead of `/api/v2/notifications/channels/testchannelid/subscriptions`

**Comment by ronanwatkins** [comment]
I'll do another sweep through it when everything's working. Looking like it's 90% there.

File: resources/sdk/purecloudjava/templates/Logger.mustache (hunk 2 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/Logger.mustache]: I corrected the log levels, seeing the error log now when I run the tests

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 3 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache]: Made some changes to the sources for the log params, let me know if any of these problems persist.  I'll do my own validation as well.

File: resources/sdk/purecloudjava/templates/Configuration.mustache (hunk 14 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/Configuration.mustache:140]: This should be fixed

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 3 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache]: Still not seeing the request body logged, working on that now.

File: resources/sdk/purecloudjava/templates/Configuration.mustache (hunk 14 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/Configuration.mustache:140]: Still have the issue with this. It's always true no matter what I do with the config

File: resources/sdk/purecloudjava/templates/Configuration.mustache [hunk 15]
```diff
@@ -20,5 +36,245 @@ public class Configuration {
...
+      return new String(hashedBytes, StandardCharsets.UTF_8);
+    } catch (Exception e) {
+      return "";
+    }
+  }
+
+  public void runConfigChecker() {
+    while (true) {
+        try {
+          Thread.sleep(30 * 1000);
```
Comment (timsmithgenesys) [inline resources/sdk/purecloudjava/templates/Configuration.mustache:135]: Have you considered using a file watcher for this instead of this blocking thread sleep? https://stackoverflow.com/questions/16251273/can-i-watch-for-single-file-change-with-watchservice-not-the-whole-directory

File: resources/sdk/purecloudjava/templates/ApiClient.mustache [hunk 16]
```diff
@@ -1178,30 +1309,107 @@ public class ApiClient implements AutoCloseable {
...
+        private String logLevel = "none";
+        private String logFilePath = null;
+        private String logFormat = "text";
+        private Boolean logToConsole = true;
+        private Boolean logRequestBody = false;
+        private Boolean logResponseBody = false;
+
+        public void setLogLevel(String logLevel) {
+            List<String> validArguments = Arrays.asList("trace", "debug", "error", "none");
+            String match = validArguments.stream()
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/ApiClient.mustache:1331]: Cool use of stream!

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 17]
```diff
@@ -168,22 +168,121 @@ Specify the connector in the builder:
...
         .withRetryConfiguration(retryConfiguration)
 ```
 Set the `maxRetryTimeSec` to the number of seconds to process retries before returning an error.
+Set the 'MaxRetryTimeSec' to the number of seconds to process retries before returning an error.
 When the retry time is a a positive integer, the SDK will follow the recommended backoff logic using the provided configuration.
 The best practices are documented in the [Rate Limiting](https://developer.mypurecloud.com/api/rest/rate_limits.html) Developer Center article.
 
+#### SDK Logging
+
+Logging of API requests and responses can be controlled by a number of parameters on the `Configuration`'s `Logger` instance.
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:191]: Need to create an instance of ApiClient.LoggingConfiguration for this SDK

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 18]
```diff
@@ -168,22 +168,121 @@ Specify the connector in the builder:
...
+Example logging configuration:
+```{"language":"java"}
+ApiClient.LoggingConfiguration loggingConfiguration = new ApiClient.LoggingConfiguration();
+loggingConfiguration.setLogLevel("trace");
+loggingConfiguration.setLogFormat("json");
+loggingConfiguration.setLogRequestBody(true);
+loggingConfiguration.setLogResponseBody(true);
+loggingConfiguration.setLogToConsole(true);
+loggingConfiguration.setLogFilePath("/var/log/javasdk.log");
+```
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:215]: Show that the APIClient must be built using `withLoggingConfiguration(loggingConfiguration)` to apply it

File: resources/sdk/purecloudjava/templates/Configuration.mustache [hunk 19]
```diff
@@ -20,5 +36,246 @@ public class Configuration {
...
+    configuration.applyConfigFromFile();
+    if (autoReloadConfig) {
+      Runnable runnable = () -> configuration.runConfigChecker();
+      Thread configCheckerThread = new Thread(runnable);
+      configCheckerThread.setDaemon(true);
+      configCheckerThread.start();
+    }
+  }
+
+  public void applyConfigFromFile() {
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/Configuration.mustache:54]: The methods in this file should be private

File: resources/sdk/purecloudjava/templates/ApiClient.mustache [hunk 20]
```diff
@@ -211,6 +276,40 @@ public class ApiClient implements AutoCloseable {
...
     }
 
+    public void setBasePath(String basePath) {
+        this.basePath = basePath;
+    }
+
+    /**
+     * Helper method to apply logging settings to logger 
+     */
+    public void applyLoggingConfiguration() {
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/ApiClient.mustache:286]: This should be private

File: resources/sdk/purecloudjava/templates/Configuration.mustache (hunk 14 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/Configuration.mustache]: I realized that autoReloadConfig was only being handled initially.   I added a couple of helper methods and this should be fixed now.

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 20 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache]: This is actually being used in the Configuration.java file, so I changed to a default scope.  It's being used when autoReloadConfig is enabled and a config file is deleted from the directory.  In this case, I designed it so that the program reverts to the logging configuration object and applies it with this method. I can remove this behavior and instead take no action on delete if needed?

File: npm-shrinkwrap.json [hunk 21]
```diff
@@ -498,7 +498,7 @@
     },
     "clone": {
       "version": "2.1.2",
-      "resolved": "https://registry.npmjs.org/clone/-/clone-2.1.2.tgz",
+      "resolved": "https://purecloud.jfrog.io/purecloud/api/npm/inin-internal-npm/clone/-/clone-2.1.2.tgz",
```
Comment (ronanwatkins) [inline npm-shrinkwrap.json:501]: It would be best to undo the changes to this file. We want to keep it using the public registry to avoid people outside of the organization having issues building.

File: resources/sdk/purecloudjava/templates/Logger.mustache [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+            if (!logResponseBody)
+                this.responseBody = null;
+            if (logFormat == LogFormat.JSON) {
+                ObjectMapper objectMapper = new ObjectMapper();
+                objectMapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss"));
+                objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+                try {
+                    return objectMapper.writeValueAsString(this);
+                } catch (Exception e) {
+                    e.printStackTrace();
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/Logger.mustache:367]: Shouldn't print this. Also, for me it's throwing an exception telling me to add the following line:
`objectMapper.disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);`
When I add that it prints an empty JSON object.
The API I was calling is `usersApi.getUsersMe(null, null);`

Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/Logger.mustache:367]: Also it seems to be hitting this code despite logFormat being set to Text and successfully logging text to the console and file

Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/Logger.mustache:367]: And it's printing text when logFormat is set to JSON

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 23]
```diff
@@ -168,22 +168,124 @@ Specify the connector in the builder:
...
 ```
 Set the `maxRetryTimeSec` to the number of seconds to process retries before returning an error.
+Set the 'MaxRetryTimeSec' to the number of seconds to process retries before returning an error.
 When the retry time is a a positive integer, the SDK will follow the recommended backoff logic using the provided configuration.
 The best practices are documented in the [Rate Limiting](https://developer.mypurecloud.com/api/rest/rate_limits.html) Developer Center article.
 
+#### SDK Logging
+
+Logging of API requests and responses can be controlled by a number of parameters on the `Configuration`'s `Logger` instance.
+Need to create an instance of ApiClient.LoggingConfiguration for this SDK.
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:192]: Please edit this line and the line above. I didn't mean for you to copy that comment in word for word. Have it show that people need to create an instance of ApiClient.LoggingConfiguration, set values on it and build the APIClient with it.

File: resources/sdk/purecloudjava/tests/.idea/misc.xml (hunk 6 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/tests/.idea/misc.xml:1]: Also .gitignore above should be deleted

File: resources/sdk/purecloudjava/tests/tests.iml (hunk 7 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/tests/tests.iml:1]: This shouldn't be added to the PR. You can add "resources/sdk/purecloudjava/tests/tests.iml" to the root .gitignore to have it ignored.

File: resources/sdk/purecloudjava/templates/ApiClient.mustache (hunk 20 above)
Comment (ronanwatkins) [reply resources/sdk/purecloudjava/templates/ApiClient.mustache]: Yeah I think removing that behavior and taking no action on delete is best

File: resources/sdk/purecloudjava/templates/README.mustache (hunk 23 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/README.mustache:192]: Oh, that is actually shown on lines 207 - 218.  Is that what you were looking for?
image

File: resources/sdk/purecloudjava/templates/Logger.mustache (hunk 22 above)
Comment (diemenschmaschine92) [reply resources/sdk/purecloudjava/templates/Logger.mustache:367]: This turned out to be caused by changing the access level of the getters to "default".  I set them back to public and seeing the logs in json format again.

File: resources/sdk/purecloudjava/templates/README.mustache [hunk 24]
```diff
@@ -168,22 +168,123 @@ Specify the connector in the builder:
...
+1. The SDK will look for `%HOMEDRIVE%%HOMEPATH%\.genesyscloudjava\config` on Windows, or `$HOME/.genesyscloudjava/config` on Unix.
+2. Provide a valid file path to the ApiClient
+
+Example setting the configuration file:
+
+```{"language": "java"}
+ApiClient apiClient = ApiClient.Builder.standard().withConfigFilePath("/path/config").build();
+```
+
+The SDK will check every 30 seconds to see if the config file has been updated, regardless of whether a config file was present at start-up. To disable this behaviour, set `autoReloadConfig` to false.
```
Comment (ronanwatkins) [inline resources/sdk/purecloudjava/templates/README.mustache:232]: Change this to show changes will be applied instantly
